doc/developer/design/20210413_source_sink_resource_sharing.md

```diff
@@ -0,0 +1,269 @@
+get if you'd pre-committed to those views. The key insight is that we can
+create a temporary backfill consumer when a new instance is instantiated:
+
+```
+          steady-state primary                    temporary backfill
+         +---------------------+                 +---------------------+
+         | librdkafka consumer |                 | librdkafka consumer |
+         +---------------------+                 +---------------------+
+         |    Avro decoding    |                 |    Avro decoding    |
+         +---------------------+                 +---------------------+
+       /                        \                          |
+      /                           \                        |
+kafka_src/1                     kafka_src/2           kafka_src/3
+    |                                |                     |
+  view1                            view2                 view3
+```
+
+Once that backfill consumer has caught up to the steady-state primary
+consumer—i.e., once the consumer offsets are identical— we can cut `kafka_src/3`
+over to the primary consumer, and jettison the backfill consumer:
+
+```
+          steady-state primary
+         +---------------------+
+         | librdkafka consumer |
+         +---------------------+
+         |    Avro decoding    |
+         +---------------------+
+       /                        \------------------------+
+      /                           \                       \
+kafka_src/1                     kafka_src/2           kafka_src/3
+    |                                |                     |
+  view1                            view2                 view3
+```
+
+### Unmaterialized PostgreSQL sources
+
+The MVP for PostgreSQL sources will not support unmaterialized sources. It's not
+fundamentally impossible to have multiple instances of a PostgreSQL source, but
+each instance of a PostgreSQL source is prohibitively expensive. Forcing
+PostgreSQL sources to be materialized is an easy way to ensure that only one
+instance of a PostgreSQL source is ever created.
+
+Roughly speaking, PostgreSQL's replication protocol works per database, while
+a PostgreSQL source in Materialize corresponds to just one table in that
+database. If you were to instantiate a PostgreSQL source twice in Materialize,
+you'd wind up streaming two separate copies of the entire database to
+Materialize, which is certain to be surprising to users.
+
+With dynamic resource sharing, creating a new instance of a PostgreSQL source
+would instead backfill the missing data for just the one table, then cut over to
+the already existing replication stream. This will be cheap enough that we'll
+be comfortable allowing PostgreSQL sources to be unmaterialized.
+
+### Discussion
+
+The major downside of this approach is the implementation complexity. Cutting
+source instances over from a backfill resource to the steady-state resource will
+require a lot of fiddly logic and testing.
```

Comment:
I think this optimization is absolutely required in the long run and worth nearly any amount of implementation complexity it might require; in fact, I think you understate its importance throughout this document. Unmaterialized Postgres sources alone are an absolute requirement IMO, and dynamic sharing would be worth it to get that alone. Saving Kafka resources is then the cherry on top.